cpp/src/io/parquet/delta_enc.cuh

```diff
@@ -164,9 +166,9 @@ class delta_binary_packer {
   }
 
   // Signed subtraction with defined wrapping behavior.
-  inline __device__ zigzag128_t subtract(zigzag128_t a, zigzag128_t b)
+  inline __device__ T subtract(T a, T b)
   {
-    return static_cast<zigzag128_t>(static_cast<uleb128_t>(a) - static_cast<uleb128_t>(b));
+    return static_cast<T>(static_cast<U>(a) - static_cast<U>(b));
```

Comment:
oh my, I don't think I understood this correctly before.
So if `a` is smaller than `b`: unsigned result "underflows" and becomes ~max_uint.
And then what happens when it's cast back to signed?

Reply:
The point of all this is to get defined behavior when we overflow or underflow. Sadly in C++ signed over/underflow has undefined behavior, but unsigned overflow is well defined. So when we're calculating deltas, we'll switch to the unsigned domain to get the defined behavior, and then cast back to signed. 1 - 2 is 0xffffffff whether signed or unsigned, and we get -1. INT_MIN - 1 is undefined in signed, but 0x7fffffff in unsigned (although every compiler I've tried yields 0x7fffffff for signed anyway).

Reply:
diabolical.